packages/ui/react-ui-form/src/components/Form/Form.tsx

```diff
@@ -128,6 +129,127 @@ export const Form = <T extends BaseObject>({
+                const typeLiteral = baseNode
+                  ? getDiscriminatedType(baseNode, values[name] as any)
+                  : findNode(elementType, AST.isTypeLiteral);
+
+                if (typeLiteral) {
+                  const schema = S.make(typeLiteral);
+
+                  return (
+                    <div key={name} role='none' className={mx(padding)}>
+                      <InputHeader>{label}</InputHeader>
+                      <div role='none' className='flex flex-col gap-1'>
+                        {arrayValues.map((value, index) => {
+                          return (
+                            <div key={index} role='none' className='flex items-center gap-1'>
+                              <div role='none' className='flex-1'>
+                                <Form<any>
+                                  schema={schema}
+                                  path={[...key, index.toString()]}
+                                  values={value}
+                                  onValuesChanged={(childValues) => {
+                                    const newValues = [...arrayValues];
+                                    newValues[index] = childValues;
+                                    inputProps.onValueChange(name, 'object', newValues);
+                                  }}
+                                  Custom={Custom as any}
+                                />
+                              </div>
+                              <IconButton
+                                icon='ph--trash--regular'
+                                iconOnly
+                                label={t('button remove')}
+                                onClick={() => {
+                                  const newValues = arrayValues.filter((_, i) => i !== index);
+                                  inputProps.onValueChange(name, type, newValues);
+                                }}
+                              />
+                            </div>
+                          );
+                        })}
+                      </div>
+                      <div role='none' className='flex justify-between items-center plb-1'>
+                        <IconButton
+                          icon='ph--plus--regular'
+                          iconOnly
+                          label={t('button add')}
+                          onClick={() => {
+                            const newValues = [...arrayValues, {}];
+                            inputProps.onValueChange(name, type, newValues);
+                          }}
+                        />
+                      </div>
+                    </div>
+                  );
+                }
+
+                return (
+                  <div key={name} role='none'>
+                    Nested form not supported in arrays, yet.
+                  </div>
+                );
+              }
+
+              return (
```

Comment:
this inline function is getting a bit long. suggest splitting it out a local function at the bottom of this file to generate the component based on the type.